Helm chart: make more fields configurable for the deployment, daemonset and storage class

Misteur-Z opened the pull request:
### Helm chart

**Is this a bug fix or adding new feature?**

Start of my work: pod annotations not being specific to the Controller deployment or the Node daemonset (bug)
==> new feature: make everything properly configurable

**What is this PR about? / Why do we need it?**

We should be able to set annotations, resources, nodeSelector, tolerations, affinity, etc. specifically. Improvement from https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/398 not fully addressing these issues.

I tried to find an elegant solution to this issue but it's very hard not to create breaking changes in the Helm values file.

**What testing is done?** 

Tested in AWS EKS 1.19.6 with Helm 3.5.3

**What else?**

I added a changelog for the Helm chart, could be useful.

Misteur-Z commented on charts/aws-efs-csi-driver/templates/controller-deployment.yaml:
```diff
@@ -19,8 +19,8 @@ spec:
-      {{- if .Values.node.podAnnotations }}
-      annotations: {{ toYaml .Values.node.podAnnotations | nindent 8 }}
+      {{- if .Values.controller.podAnnotations }}
+      annotations: {{ toYaml .Values.controller.podAnnotations | nindent 8 }}
```
`.Values.podAnnotations` was actually unused => not working (found out while using `kube2iam`)

Misteur-Z commented:
I signed it

wongma7 commented:
/ok-to-test
/check-cla

wongma7 commented:
Heads up, you might have to do some rebasing, I might merge a bunch of 1.2.x changes first in the next few days, however in principle I approve of this PR.

Misteur-Z commented:
I signed it

Misteur-Z commented:
* Seems good to go, let me test the Chart again just in case
* I didn't update the Helm chart changelog all the way to version `1.0.0`

Misteur-Z commented:
I signed it

Misteur-Z commented:
* Seems good to go, tested the Chart again in EKS
* I didn't update the Helm chart changelog all the way to version `1.0.0`

wongma7 commented on charts/aws-efs-csi-driver/Chart.yaml:
```diff
@@ -2,8 +2,8 @@ apiVersion: v1
 description: A Helm chart for AWS EFS CSI Driver
-version: 1.2.2
-kubeVersion: ">=1.17.0-0"
+version: 1.3.0
```
do we need to bump the major version? I'm not really sure about best practice but I think yes?

wongma7 replied:
In the apst, we have bumped minor version to signal that a new efs driver version is out.

So e.g. currently it so happens that 1.2.x correspondx to efs driver 1.2.y.

That's why I think we need a major version bump

Misteur-Z replied:
I am creating breaking changes in the values file, that's why I thought about bumping the major version. I would say most helm charts versions don't match the app version.

Misteur-Z replied:
Good idea since there is a lot of potential breaking changes! Let's bump to `2.0.0` to make it significant.

wongma7 commented on charts/aws-efs-csi-driver/Chart.yaml:
```diff
@@ -2,8 +2,8 @@ apiVersion: v1
-version: 1.2.2
-kubeVersion: ">=1.17.0-0"
+version: 1.3.0
+kubeVersion: ">=1.18.0-0"
```
what change is requiring 1.18?

wongma7 replied:
oh never mind, I see it's because of v1 csidriver in the changelog.

This complicates matters because technically the driver works with 1.17+ but now the primary install method requires 1.18+.

looks like ebs did the same change though so...https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/836

krmichel replied:
They did but I don't think it was handled correctly there.  They didn't change the kube version dependency in the chart.  I don't think it requires that the chart only work on 1.18+

Misteur-Z commented on charts/aws-efs-csi-driver/templates/_helpers.tpl:
```diff
@@ -49,9 +49,9 @@ Create the name of the service account to use
+    {{ default (include "aws-efs-csi-driver.fullname" .) .Values.controller.serviceAccount.name }}
 {{- else -}}
-    {{ default "default" .Values.serviceAccount.controller.name }}
+    {{ default "default" .Values.controller.serviceAccount.name }}
```
`serviceAccount` is controller specific

Misteur-Z commented:
Merged master again:
- fixed `logLevel` not being specific to either the controller deployment or node daemonset
- moved `serviceAccount` into `controller` since it's specific to its deployment

wongma7 commented:
thanks, everything lgtm but I'm just hesitant about the versioning, overall I'm satisfied that this PR is making the chart more consistent and clearly partitioning values along controller/node lines.

AFter it merges, I won't make the release immediately, I need to test especially the service account and RBAC wiring, since we've had longstanding issues there: https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/347 https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/421

wongma7 commented:
@kbasv @ayberk  could you guys help me audit the helm chart and suggest any changes? This PR is making some breaking changes and bumping to 2.0.0 so as @Misteur-Z said, if there's anything else we want to change, now is the time to do it!

wongma7 commented:
Also @krmichel if you wouldn't mind reviewing, yo uare a helm chart maintainer so would appreciate your eyes on this!!

ayberk commented on charts/aws-efs-csi-driver/CHANGELOG.md:
```diff
@@ -0,0 +1,15 @@
+* Multiple breaking changes in values file
+* Support only Kubernetes `>=1.18` with `CSIDriver` API version `storage.k8s.io/v1`
+
+## New features
```
Can we have a specific section for "Breaking Changes"? Ideally we don't want the users to go read the template and figure out every single change they need to make.

Misteur-Z replied:
Tried to explain breaking changes in b1c989de876b5cc719153f51dbe3a831a0a731ed

ayberk commented on charts/aws-efs-csi-driver/templates/controller-deployment.yaml:
```diff
@@ -32,29 +32,29 @@ spec:
             privileged: true
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
-          imagePullPolicy: IfNotPresent
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
```
Can we move `pullPolicy` to their own fields under `node` and `controller`? Just to be consistent.

wongma7 replied:
that would be less consistent cuz it would break the pattern of image structs all having the same fields pullPolicy, tag , & repository. The top-level `image` is the driver image which should be the same for both node and controller.

wongma7 commented on charts/aws-efs-csi-driver/values.yaml:
```diff
@@ -10,85 +13,99 @@ image:
+
+node:
+  # Number for the log level verbosity
+  logLevel: 5
```
while we r here, default should be lower, like 2: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/854

Misteur-Z replied:
Lowered to 2 in b1c989de876b5cc719153f51dbe3a831a0a731ed

krmichel commented on charts/aws-efs-csi-driver/templates/csidriver.yaml:
```diff
@@ -1,4 +1,4 @@
-apiVersion: storage.k8s.io/v1beta1
+apiVersion: storage.k8s.io/v1
```
This would be better handled like this and then you wouldn't need to increase the minimum kube version.
https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/856/files#diff-e167822724b087c64ed35c6f75c8af5d054a981e00b59488145796c69c6c611fR1

`apiVersion: {{ ternary "storage.k8s.io/v1" "storage.k8s.io/v1beta1" (semverCompare ">=1.18.0-0" .Capabilities.KubeVersion.Version) }}`

Misteur-Z replied:
Fixed in b1c989de876b5cc719153f51dbe3a831a0a731ed

krmichel reviewed (commented):
> Also @krmichel if you wouldn't mind reviewing, yo uare a helm chart maintainer so would appreciate your eyes on this!!

@wongma7 I will take a look at this.  I have started a review but don't have time to finish it right now.  Please don't merge until I have had a chance to finish my review.  There are a number of breaking changes that could be avoided if the changes were made in two steps.  Is that a concern that I should make comments about during my review or should I just assume that breaking changes are fine?

Misteur-Z commented:
Merged master, addressed all the comments.

krmichel commented on charts/aws-efs-csi-driver/values.yaml:
```diff
@@ -10,85 +13,99 @@ image:
+    #   nameservers:
+    #     - 169.254.169.253
+  podAnnotations: {}
+  resources:
```
move `{}` to this line to make the example easier to read

krmichel commented on charts/aws-efs-csi-driver/values.yaml:
```diff
@@ -10,85 +13,99 @@ image:
   # path on efs when deleteing an access point
   deleteAccessPointRootDir: false
+  podAnnotations: {}
+  resources:
```
move `{}` to this line to make the example easier to read

krmichel commented on charts/aws-efs-csi-driver/values.yaml:
```diff
@@ -10,85 +13,99 @@ image:
+node:
+  # Number for the log level verbosity
+  logLevel: 2
+  hostAliases:
```
move `{}` to this line to make the example easier to read

krmichel commented on charts/aws-efs-csi-driver/values.yaml:
```diff
@@ -10,85 +13,99 @@ image:
+    #   ip: 10.10.2.2
+    #   region: us-east-2
+  dnsPolicy: ""
+  dnsConfig:
```
move `{}` to this line to make the example easier to read

krmichel commented on charts/aws-efs-csi-driver/CHANGELOG.md:
```diff
@@ -0,0 +1,69 @@
+
+# v2.0.0
+
+## Breaking changes
```
You don't mention moving nodeSelector and resources under node.  It is probably also worth noting that you fixed a bug where the controller was using podAnnotations and tolerations from the node section and is now using the correct values from the controller section.  It might be worth putting a section in the readme under the helm instructions that point here for upgrading from versions prior to 2.0.0

Misteur-Z replied:
Tried my best to clarify everything, tell me if it's not enough

krmichel commented on charts/aws-efs-csi-driver/templates/controller-deployment.yaml:
```diff
@@ -32,29 +32,29 @@ spec:
         {{- . | toYaml | nindent 8 }}
         {{- end }}
       serviceAccountName: {{ include "aws-efs-csi-driver.serviceAccountName" . }}
       priorityClassName: system-cluster-critical
```
It might be worth considering adding the ability to set topologySpreadConstraints while this change is being made.  This is untested but might look something like this.

```
      {{- with .Values.controller.topologySpreadConstraints }}
      {{- $tscLabelSelector := dict "labelSelector" (dict "matchLabels" (dict "app" "efs-csi-controller")) }}
      topologySpreadConstraints:
        {{- range . }}
        - {{ mergeOverwrite . $tscLabelSelector | toJson }}
        {{- end }}
      {{- end }}
```

Misteur-Z replied:
I am not sure of the implications of changing anything here, are we supporting only Kubernetes running on Linux hosts?

Misteur-Z replied:
Not adding `topologySpreadConstraints` for now, sticking with the current `nodeSelector` and `priorityClassName`.

wongma7 replied:
works for me, I don't think topologySpreadConstraints would make sense on node, we can always add later to controller if somebody asks for it.

krmichel reviewed (changes requested):
Okay, I finished a pass through the whole thing.

krmichel commented on charts/aws-efs-csi-driver/Chart.yaml:
```diff
@@ -1,9 +1,9 @@
 apiVersion: v1
```
While making a breaking change might as well make this version helm API v2.

wongma7 commented:
/hold

Misteur-Z commented on charts/aws-efs-csi-driver/templates/controller-deployment.yaml:
```diff
@@ -32,29 +32,29 @@ spec:
       serviceAccountName: {{ include "aws-efs-csi-driver.serviceAccountName" . }}
       priorityClassName: system-cluster-critical
       tolerations:
         - operator: Exists
```
I wanted to fix this but I didn't want to introduce some breaking change or new issue. Can we let users set `tolerations`?

Misteur-Z replied:
Switching to let the user define the whole `tolerations` array in values file.

Misteur-Z commented on charts/aws-efs-csi-driver/templates/node-daemonset.yaml:
```diff
@@ -50,9 +50,7 @@ spec:
-      {{- if .Values.node.dnsPolicy }}
-      dnsPolicy: "{{ .Values.node.dnsPolicy }}"
-      {{- end }}
+      dnsPolicy: {{ .Values.node.dnsPolicy }}
```
Test the helm chart this morning, feeding `dnsPolicy` with an empty value will use default `ClusterFirst`

Misteur-Z commented on deploy/kubernetes/base/controller-serviceaccount.yaml:
```diff
@@ -47,7 +47,7 @@ metadata:
   - kind: ServiceAccount
     name: efs-csi-controller-sa
-    namespace: sample
+    namespace: default
```
Not related to the helm chart

Misteur-Z commented:
Feels good to go @krmichel, not sure if there is anything to fix in `deploy/` or elsewhere @wongma7

krmichel commented on charts/aws-efs-csi-driver/templates/node-daemonset.yaml:
```diff
@@ -50,28 +50,25 @@ spec:
       {{- end }}
       priorityClassName: system-node-critical
-      tolerations:
-        - operator: Exists
```
The daemonset and the deployment should behave differently.  Removing the operator: Exists from the daemonset I think would be surprising to users.  I think a daemonset needs some sort of default tolerations since it is supposed to be running on "every" node.  I think a reasonable solution might be to have the operator: Exists as a default in the values.yaml and then if a user wants they have a method of getting rid of that toleration.

```
node:
  ...
  tolerations:
    - operator: Exists
```

Misteur-Z replied:
Fixed: https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/406/commits/a2aa0794e2ddd4921e068fc044d201a67e86b662

krmichel commented:
> Feels good to go @krmichel, not sure if there is anything to fix in `deploy/` or elsewhere @wongma7

I think it looks pretty good except dealing with the daemonset tolerations.  Once that is addressed I should be good with it.

krmichel reviewed (approved):
looks good

wongma7 commented:
/hold cancel

Sorry Ive been quite busy for last 2 weeks but will do final pass and merge soon

wongma7 commented:
/hold cancel

Sorry Ive been quite busy for last 2 weeks but will do final pass and merge soon

wongma7 commented:
/approve

thank you very much. Appreciate the patience, reviewer bandwidth is quite low... do you mind squashing the commits into one or two? using git rebase -i or something similar. It will make it easier to backport to release-1.2 branch  ( which to be fair I am not 100% sure we should do yet). Merging this PR WON'T immediately release the 2.0.0 chart, somebody still needs to do some manual stuff to release the chart.

Misteur-Z commented:
Squashed into a single commit

wongma7 commented:
/lgtm

rmgpinto commented:
@wongma7 can you release the 2.0.0 chart?
Thanks

wongma7 commented:
@rmgpinto it's coming https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/458
